### PR TITLE
Fix cdb_sequence_nextval_qe if connection is terminated

### PR DESCRIPTION
**Pull request by gfphoenix78**
Fix an assert failure in cdb_sequence_nextval_qe().
When QE wants to wait a message from QD to allocate the next value,
it receives a message with type 'X', which QD notifies QE to terminate.
Message with type 'X' is valid. So, cdb_sequence_nextval_qe should
handle it correctly.

No test for this commit. Because we need to register a fault injector in
cdb_sequence_nextval_qe() to suspend the execution of the QE process,
and later we want to resume the execution of cdb_sequence_nextval_qe()
by reset the registered fault inject. It works in normal cases. But we
need to kill the postmaster of the blocked QE process between suspending
the QE and resuming the QE. Resuming of the blocked QE failed because of
the postmaster of the QE is dead.

Co-authored-by: Paul Guo <pguo@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kainwen** [comment]
Would you please add more comments in the code?

Also, if no test cases added, please explain clearly in the code and the PR message.

Thanks!

**Comment by gfphoenix78** [comment]
Ok, add the reason why the commit has no test case.

**Comment by gfphoenix78** [comment]
Why does this PR have no test case?

No test for this commit. Because we need to register a fault injector in
cdb_sequence_nextval_qe() to suspend the execution of the QE process,
and later we want to resume the execution of cdb_sequence_nextval_qe()
by reset the registered fault inject. It works in normal cases. But we
need to kill the postmaster of the blocked QE process between suspending
the QE and resuming the QE. Resuming of the blocked QE failed because of
the postmaster of the QE is dead.

**Comment by gfphoenix78** [comment]
How to reproduce?
* psql side:
```SQL
CREATE TABLE tmp_table_cache (a int);
INSERT INTO tmp_table_cache VALUES (2),(1),(5);
CREATE SEQUENCE tmp_seq INCREMENT 1 MINVALUE 1 MAXVALUE 40 START 1 CACHE 3 NO CYCLE;
SELECT nextval('tmp_seq'), * FROM tmp_table_cache;
```
* Before running the last SQL, attach two gdbs to the QD process and QE of seg1, and break `checkSegmentAlive()` in gdb of QD console, break `cdb_sequcence_nextval_qe()` in gdb of the QE console.
* Run the last SQL in step 1, and kill the postmaster process of the attached QE process.
* Wait a moment for the QD to check for the expected error.
* Print the call stack of the QD stack.

**Comment by asimrp** [comment]
The change looks good.  One question regarding the repro.  Do we have to kill the postmaster of QE?  What would happen if the query is canceled or terminated using `pg_cancel/terminate_backend()` on the QD?

**Comment by gfphoenix78** [comment]
Good question. I run `SELECT nextval('tmp_seq'), * FROM tmp_table_cache;` and the QE is blocked on the faultinjector. Then I open a new terminal to run pg_terminate_backend(<pid of the QD of the above query>). The result is:
```
d0=# SELECT nextval('tmp_seq'), * FROM tmp_table_cache;
FATAL:  terminating connection due to administrator command
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

**Comment by asimrp** [comment]
That means QD crashed.  What do the QD logs say? Was a core dumped?

**Comment by gfphoenix78** [comment]
Yes, I send a signal to the QD by pg_terminate_backend(). It is what I expect. The log is:
```
2019-11-01 02:00:59.525764 UTC,"gpadmin","d0",p20851,th-1788467072,"[local]",,2019-11-01 02:00:49 UTC,0,con52,cmd1,seg-1,,dx2684,,sx1,"LOG","00000","statement: SELECT nextval('tmp_seq'), * FROM tmp_table_cache;",,,,,,"SELECT nextval('tmp_seq'), * FROM tmp_table_cache;",0,,"postgres.c",1617,
2019-11-01 02:02:15.755594 UTC,"gpadmin","d0",p20809,th-1788467072,"[local]",,2019-11-01 01:59:08 UTC,0,con51,cmd10,seg-1,,dx2687,,sx1,"LOG","00000","statement: select pg_terminate_backend(20851);",,,,,,"select pg_terminate_backend(20851);",0,,"postgres.c",1617,
2019-11-01 02:02:15.967038 UTC,"gpadmin","d0",p20851,th-1788467072,"[local]",,2019-11-01 02:00:49 UTC,770,con52,cmd2,seg-1,,dx2684,x770,sx1,"FATAL","57P01","terminating connection due to administrator command",,,,,,"SELECT nextval('tmp_seq'), * FROM tmp_table_cache;",0,,"postgres.c",3750,
2019-11-01 02:02:15.973361 UTC,"gpadmin","d0",p20851,th-1788467072,"[local]",,2019-11-01 02:00:49 UTC,770,con53,,seg-1,,dx2684,x770,sx1,"LOG","00000","The previous session was reset because its gang was disconnected (session id = 52). The new session id = 53",,,,,,,0,,"cdbgang.c",713,
```
This action doesn't make QD send a terminate message to QE.

File: src/backend/commands/sequence.c [hunk 1]
```diff
@@ -1883,9 +1883,14 @@ cdb_sequence_nextval_qe(Relation	seqrel,
 		if (retval == EOF)
 			ereport(ERROR,
 					(errcode(ERRCODE_INTERNAL_ERROR),
-					 errmsg("nextval: QD closed the connection")));
+					 errmsg("nextval: connection is gone unexpectedly!")));
```
Comment (pivotal-ning-yu) [inline src/backend/commands/sequence.c]: Let's do not add a `!` in the error message.

File: src/backend/commands/sequence.c [hunk 2]
```diff
@@ -1883,9 +1883,14 @@ cdb_sequence_nextval_qe(Relation	seqrel,
...
 	} while (retval != 1);
-	Assert(qtype == SEQ_NEXTVAL_QUERY_RESPONSE);
+	if (qtype == 'X')
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("nextval: QD closed the connection")));
+	if (qtype != SEQ_NEXTVAL_QUERY_RESPONSE)
+		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("unexpected message type='%c'", qtype)));
```
Comment (pivotal-ning-yu) [inline src/backend/commands/sequence.c]: Please add a `nextval: ` prefix to the error message.

Comment (gfphoenix78) [reply src/backend/commands/sequence.c]: OK

File: src/backend/commands/sequence.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/commands/sequence.c]: Thanks.

File: src/backend/commands/sequence.c [hunk 3]
```diff
@@ -1883,18 +1883,18 @@ cdb_sequence_nextval_qe(Relation	seqrel,
 		if (retval == EOF)
 			ereport(ERROR,
 					(errcode(ERRCODE_INTERNAL_ERROR),
-					 errmsg("nextval: connection is gone unexpectedly!")));
+					 errmsg("nextval: connection is gone unexpectedly")));
 	} while (retval != 1);
 	if (qtype == 'X')
 		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
-						errmsg("nextval: QD closed the connection")));
+				errmsg("nextval: QD closed the connection")));
```
Comment (pivotal-ning-yu) [inline src/backend/commands/sequence.c]: The indentation is bad.  If you use vim please `set ts=4 sw=4 noet cino=(0` then use `==` to auto adjust the indentation.

Comment (gfphoenix78) [reply src/backend/commands/sequence.c]: Magic.

**Pull request merged**